### ISSUE TITLE
Exception occurred in retry method that was not classified as transient

### ISSUE THREAD

**Issue by lamstutz** [firebase/firebase-functions#536]
### Related issues

https://github.com/firebase/firebase-functions/issues/522

### [REQUIRED] Version info

```
  "dependencies": {
    "@google-cloud/firestore": "^2.2.1",
    "firebase-admin": "^8.2.0",
    "firebase-functions": "^3.0.1",
  },
  "engines": {
    "node": "8"
  }
```

**node:** 8

**firebase-functions:** 3.0.1

**firebase-tools:** 7.0.1

**firebase-admin:** 8.2.0

### Steps to reproduce

```
import * as admin from 'firebase-admin';

admin.initializeApp();
const db = admin.firestore();
db.settings({ timestampsInSnapshots: true });


const users = db.collection('users');

users.doc('myUserId').update({ fieldToUpdate: 'newValue' })
```

Update method throw this error :

> { Error
>     at Http2CallStream.call.on (/srv/node_modules/@grpc/grpc-js/build/src/client.js:101:45)
>     at emitOne (events.js:121:20)
>     at Http2CallStream.emit (events.js:211:7)
>     at process.nextTick (/srv/node_modules/@grpc/grpc-js/build/src/call-stream.js:71:22)
>     at _combinedTickCallback (internal/process/next_tick.js:132:7)
>     at process._tickDomainCallback (internal/process/next_tick.js:219:9)
>   code: 13,
>   details: '',
>   metadata: Metadata { options: undefined, internalRepr: Map {} },
>   note: 'Exception occurred in retry method that was not classified as transient' }

### Were you able to successfully deploy your functions?

the deployment displays no errors

**Comment by thechenky** [comment]
Hi @lamstutz does this occur on every function invocation? Would you mind pasting more logs? Also, does this happen when updating other collections?

**Comment by audkar** [comment]
Having same error
> does this occur on every function invocation?

No. [following statement is might be wrong] this error happens only after idle function is invoked after longer period of time.

**Comment by thechenky** [comment]
From the stack trace and the fact that it happens on idle function it looks like a grpc related error that is happening on a cold start. Does this happen on other collection updates? Are you properly handling promises in this function? Because the stack trace mentions retries, it may be happening because the value you're trying to write into Firestore is somehow problematic - are you able to replicate this with another collection with a very simple interface - something like updating one field that is string?

**Comment by lamstutz** [comment]
The error occurs regularly but not systematically. Only with "update" function (whether with a simple object or an increment). Only with "triggers", not with a http function. Promises are well used and catched.

ex:

```
const categoryIds = ['a','b'];
  return users.doc(userId).update({ categoryIds });
```

Or

`return statistics.doc('users').update({ users: gcFieldValue.increment(-1) });`

**Comment by thechenky** [comment]
Hmm, it seems that something is going wrong with that particular Firestore update call - @schmidt-sebastian @hiranya911 I wonder if you have ideas on what could be going wrong here?

**Comment by bottleneck-admin** [comment]
I am having the exact same problem

**Comment by spoxies** [comment]
I have to add that in my experience the problem is also present with 'http' calls and not only with 'triggers'. It seems it is present when an instance/function went to sleep and has indeed a 'cold start'.

**Comment by stshelton** [comment]
I'm having the same problem. I'm receiving this error.

> Error
    at Http2CallStream.call.on (/srv/node_modules/@grpc/grpc-js/build/src/client.js:96:45)
    at emitOne (events.js:121:20)
    at Http2CallStream.emit (events.js:211:7)
    at process.nextTick (/srv/node_modules/@grpc/grpc-js/build/src/call-stream.js:71:22)
    at _combinedTickCallback (internal/process/next_tick.js:132:7)
    at process._tickDomainCallback (internal/process/next_tick.js:219:9

Mine seems to happen randomly in onCreate and onWrite functions. These functions with this error are triggered daily and the error has only occurred once, I've had them fire multiple times after error occurred and error has yet to return. These errors start appearing once I updated firebase functions from version 2.3.0 to 3.0.1 and firebase admin from 7.0.0 to 8.0.0

**Comment by jaycosaur** [comment]
Just thought I'd give some input here. My first instance of this error was on the 15th of July and I now get it regularly (but not consistently) across all our functions.

We having a logging system implemented on our functions that essentially tells us when a function is cold started or not, (we ping them to keep warm every minute). Prior to the 15th of July (from 2017 to now! so I have a lot of logs on this) (ie when these errors started happening to me) cloud functions would delete themselves at approx. 3-5 minute intervals from first creation, making the next invocation a cold start. Since the 15th of July this has increased substantially to greater than 5 hours(!!!) and we have seen today a function stay warm for 28 hours (causing a lot of issues to our caching). My guess would be that a previously short running connection is now having to cope with these much much much longer keep-warms.

Now unfortunately we do not ping over the weekends, for cost reduction reasons, but on the 12th (and for the last 12+ months) it was cold starting every 3-5 minutes, and on the 15th it now doesn't cold start for 5+ hours. If this is a new 'feature' of cloud functions it is amazing btw! Almost makes them never have to hit cold starts if the keep warm invocations are done right.

**Comment by bottleneck-admin** [comment]
I am getting it with Pub/Sub functions. I also think it is related to cold starts

**Comment by damienix** [comment]
Also started getting these recently :(

**Comment by schmidt-sebastian** [comment]
Sorry for all the trouble this is causing! While we are currently looking into this, we don't have a strong lead as to what is going on. Please do bear with us.

**Comment by schmidt-sebastian** [comment]
@damienix, @bottleneck-admin, @jaycosaur, @spoxies, @lamstutz:

Would you mind sending your project IDs and an approximate time window for these errors (including your timezone) to samstern@google.com? Thanks

**Comment by steren** [comment]
This also affects Cloud Run.

For googlers on this thread, 138705198 is the internal issue

**Comment by bottleneck-admin** [comment]
How do I private message you @schmidt-sebastian

**Comment by schmidt-sebastian** [comment]
Thanks for sending us your project info. Our backend team will look into the errors. While they do, 
can you quickly confirm where your GCF instances are all running in Europe and where your Firestore project is located.

Thanks!

@bottleneck-admin If you need to send project-specific or confidential data to us for issue triage, the recommended way is to open a Support ticket via https://support.google.com/

**Comment by bottleneck-admin** [comment]
Mine are in europe-west2 and project is -> Google Cloud Platform (GCP) resource location
europe-west2

**Comment by ltomes** [comment]
@steren @schmidt-sebastian 

I came across this issue doing a google search.

I am seeing the same behavior surface from GRPC calls made by `@grpc/grpc-js` which is a dependency of `@google-cloud/storage` in my case. 

We see intermittent failures in k8s pods when a large number of files are being streamed to
storage objects via @google-cloud/storage.

If those logs might be useful let me know and I can provide a project id.

**Comment by jaycosaur** [comment]
Our project is us-central1 for functions and project.

**Comment by schmidt-sebastian** [comment]
Our backend team believes that they know what the root cause is, but it might take quite a while for the issue to be fixed in all production environments.

**Comment by damienix** [comment]
Any progress on that? This is a really severe issue ;/

As per docs https://firebase.google.com/docs/functions/retries

> Cloud Functions guarantees at-least-once execution of a background function for each event emitted by an event source.

Which is no longer true. On my backend, this leads to more and more inconsistent data, as I'm getting random errors from triggered functions that would normally run without any problems :(

Has anyone tried to enable retries of a function to defend from this error, will it work for system-level errors?

**Comment by schmidt-sebastian** [comment]
With errors like these, your request will likely succeed if you retry. Our client only retries in a couple of cases where we know it is safe (we can always retry a get() request, but we cannot retry writes as we don't know whether there are any side effects). If you know that you can always retry (based on your data model), then I would recommend that as a solution.

You could also wrap your writes in a transaction, which the client retries.

**Comment by shaunwarman** [comment]
Seeing this a lot more recently with what looks to be cloud datastore. This is what I see with a trace list. It does seem quite intermittent and most common after a time of low traffic which makes me assume stale connection or some cold start related issue. I'll add some retry logic around the datastore client.

Will also do some testing by dropping from `@google-cloud/datastore@^4` back to `@google-cloud/datastore@^3`.

```

Label	                 Value
/error/message	         read ECONNRESET
/error/name	         Error
/http/response/size	 0
/http/url	         https://datastore.googleapis.com/google.datastore.v1.Datastore/RunQuery

Callstack
<anonymous function> (/node_modules/@google-cloud/trace-agent/build/src/plugins/plugin-http2.js:63:42)
startCallStream [as startCallStream] (/node_modules/google-gax/node_modules/@grpc/grpc-js/build/src/subchannel.js:105:42)
finalMetadata.then.metadataValue (/node_modules/google-gax/node_modules/@grpc/grpc-js/build/src/channel.js:205:28)
<anonymous function>
_tickDomainCallback [as _tickDomainCallback] (internal/process/next_tick.js:229:7)

**Comment by iamnels1** [comment]
Hi,
Same problem here with an onUpdate function; the problem occur randomly...

at Http2CallStream.call.on (/srv/node_modules/@grpc/grpc-js/build/src/client.js:101:45)
    at emitOne (events.js:121:20)
    at Http2CallStream.emit (events.js:211:7)
    at process.nextTick (/srv/node_modules/@grpc/grpc-js/build/src/call-stream.js:71:22)
    at _combinedTickCallback (internal/process/next_tick.js:132:7)
    at process._tickDomainCallback (internal/process/next_tick.js:219:9)

**Comment by schankam** [comment]
We're having the same problem on our side. Location is in Japan, on asia-northeast1.

We're experiencing the issue randomly on a function triggered by the cloud scheduler. Most of the time it works fine, but sometimes we're having this issue. It is pretty annoying in our case, as we need to make sure there is a check done every 30 minutes in some entries present in our database by that time. Any insights about this issue would be appreciated. I've also written to the firebase support, waiting to see what they will say.

**Comment by sodapay-chris** [comment]
I've seen this the last two days in functions triggered on asia-northeast1.
First execution fails and then all subsequent executions run correctly.

**Comment by wflores96** [comment]
I'm also experiencing this issue. I have a cloud function triggered by document create, that returns a transaction to write to a different collection. While testing the function after long periods of time (cold start) the function will fail with the errors in the original post. Subsequent attempts to trigger these functions succeed with no error. 

Is there a timeline on this @schmidt-sebastian ? This is a pretty severe issue...

**Comment by dlihhats** [comment]
I see the same problem on cold start then all fine on subsequent calls. `europe-west-1` region
I started encountering this problem on 13th August 2019.

Really looking forward to a fix *pray*

**Comment by mikron123** [comment]
Got same message today over a promise that failed.

**Comment by sambecker** [comment]
Same here.

It's incredibly difficult to diagnose a problem with no error handling and unpredictable misfires. As far as I can tell, this only happens with my `onCreate` and `onUpdate` firestore triggers which return `Promise.all` and the end of each invocation. It seems under these circumstances all promises fail (but wouldn't most of the time).

```
"firebase-admin": "^8.3.0"
"firebase-functions": "^3.2.0" 
```

Does anyone have any tips for getting to the actual underlying error, instead of `Exception occurred in retry method that was not classified as transient`?

**Comment by sambecker** [comment]
Getting these almost every day now @schmidt-sebastian is there still a plan to resolve this? I'm in `us-central1`. Is it worth downgrading `firebase-admin` to version `7` and `firebase-functions` to version `2`?

**Comment by bottleneck-admin** [comment]
If you keep your function hot using the chron scheduler then you won't get this error anymore.
Its easy if you are using a HTTP function or pub/sub. For http, use chron scheduler to send a DELETE call to the endpoint. Check the method of the incoming request and if its delete just return and end the function. I am waking my function every 1 minute.

**Comment by bottleneck-admin** [comment]
I am doing this on one endpoint that is directly measuring usage that I bill someone for so its important!

**Comment by sambecker** [comment]
@bottleneck-admin what do you mean send a delete call? Like try to delete a document that doesn't exist every minute?

**Comment by bottleneck-admin** [comment]
Sorry I should have been more clear. A DELETE method call using http. You can use any other method you want. 

```javascript
exports.recordHit = region(REGION).https.onRequest(async (req, res) => {
    if (req.method === 'DELETE') {
        console.log('I am awake');
        return res.send(200);
    }
  doOtherStuff()
});
```

**Comment by bottleneck-admin** [comment]
![Screenshot 2019-08-20 at 15 47 59](https://user-images.githubusercontent.com/53227586/63357788-ebfd4980-c361-11e9-9763-99ccc8d8a031.png)

**Comment by sambecker** [comment]
Thanks @bottleneck-admin and in order to keep firestore `onCreate`/`onWrite` triggers warm, should I be creating/updating/destroying a document every time that gets called?

**Comment by bottleneck-admin** [comment]
You could create a http function that updates the same document in that store every minute. So create a cloud scheduler job like I have above that triggers the http function which in turn triggers your db. I don't know if that will solve your problem but its worth a shot. That would be max 1440 function calls every day. 60*24 = 1440

**Comment by bottleneck-admin** [comment]
```javascript
exports.keepAlive = region('your-region').https.onRequest((req, res) => {
    doYourDbThingHere()
    return;
})
```

**Comment by sambecker** [comment]
@bottleneck-admin could I also handle this in a schedule job alongside my other firestore functions? Would this be sufficient to keep my triggers warm?

```
exports.keepFirestoreOnUpdateAlive = functions.pubsub.schedule('every 1 minute')
  .onRun(async context => {
    return admin.firestore().collection('documents').doc('document).update({field: 'data'})
  })

exports.keepFirestoreOnCreateAlive = functions.pubsub.schedule('every 1 minute')
  .onRun(async context => {
    return admin.firestore().collection('documents').add({empty: 'document'})
  })
```

**Comment by mpalmerlee** [comment]
Keeping a single instance of your function warm will only reduce the chances of getting the issue, if you get multiple simultaneous requests, firebase will spin up new instances to run your function and you will get a cold start.

Also, writing new functions that are kept warm will not work AFAIK since each function gets its' own runtime environment / instance.

**Comment by bottleneck-admin** [comment]
Sure you might as well try it. At least it might help reduce the error rates.

**Comment by sambecker** [comment]
@mpalmerlee am I understanding correctly that on some cold starts firestore triggers seem to be failing completely and irrevocably? And this has been going on for about a month? Outside of this thread does Google have a POV on this? It feels like a monster data integrity problem. cc @schmidt-sebastian

**Comment by mpalmerlee** [comment]
@sambecker That seems to be the issue from the details I've seen in this thread, I'm just watching the thread and hoping for a resolution also since I've seen the issue crop up in my projects as well.

**Comment by schmidt-sebastian** [comment]
We are still trying to isolate the root cause. We do hear you loud and clear though. Sorry for the trouble and sluggish responses.

**Comment by sambecker** [comment]
Thanks for the response @schmidt-sebastian would it be helpful to (confidentially) share my project details? Is this happening to all firebase projects or just specific ones?

**Comment by schmidt-sebastian** [comment]
Our current theory (well mostly just mine)  is that this might be related to changes in our networking layer as we switched from a GRPC binary to a pure JavaScript implementation. If you can, do you mind downgrading to firebase-admin 7 and let us know if this issue continues to reproduce? firebase-admin uses `@google/cloud-firestore` 1.x which relies on binary GRPC layer.

**Comment by sambecker** [comment]
@schmidt-sebastian if I downgrade to `firebase-admin v7` do I also need to downgrade to `firebase-functions v2`?

**Comment by schmidt-sebastian** [comment]
@sambecker functions v3 will pull in firebase-admin v8 even if you manually specify an additional dependency on firebase-admin v7. So yes, you likely have to downgrade.

**Comment by sambecker** [comment]
@schmidt-sebastian just downgraded my dependencies. We'll see what happens. In the meantime, can someone look at my project to see if there's anything else I can do?

**Comment by schmidt-sebastian** [comment]
As of now, we don't believe that this is a project specific issue. Our current theory is that this is related to how the GRPC library handles connections that have become unresponsive.

**Comment by sambecker** [comment]
That makes sense @schmidt-sebastian and in this case is it a cold start that is deemed too slow to start up? Is it a firestore/db connection that’s too slow? Am I right in thinking it has nothing to do with the logic or dependencies of the specific function/trigger?

**Comment by schmidt-sebastian** [comment]
When we saw similar issues in the past, it was related to a "cold start" of the network stack. We have yet to confirm whether this is a similar issue. It is very likely unrelated to the logic in you function. One change I can recommend is that you should share you Firestore instance between invocations (move it to the global scope) if you don't already do that.

**Comment by sambecker** [comment]
@schmidt-sebastian I'm not sure. What are the mechanics/syntax of sharing one Firestore instance through the global scope? Does that mean not calling `admin.firestore()` from different functions but storing it once in a `const`? I've done my best to split up my code into bite-sized so it's pretty modular. Is that bad?

**Comment by schmidt-sebastian** [comment]
See here: https://cloud.google.com/functions/docs/bestpractices/tips#use_global_variables_to_reuse_objects_in_future_invocations

With that said, if you are using `admin.firestore()` as your entry point, then you will already use Firestore from an essentially global scope since we cache the instance for you. We recommend that you re-use an existing Firestore instances (either by keeping a reference manually or by calling `admin.firestore()`) so that all Firestore operations share the same GRPC channel.

**Comment by florianlenz** [comment]
I am not sure if it's related, but i am getting a similar error (that is at least my impression). 

```
{ Error: Missing or insufficient permissions.
    at Http2CallStream.call.on (/srv/node_modules/@grpc/grpc-js/build/src/client.js:96:45)
    at emitOne (events.js:121:20)
    at Http2CallStream.emit (events.js:211:7)
    at process.nextTick (/srv/node_modules/@grpc/grpc-js/build/src/call-stream.js:71:22)
    at _combinedTickCallback (internal/process/next_tick.js:132:7)
    at process._tickDomainCallback (internal/process/next_tick.js:219:9)
  code: 7,
  details: 'Missing or insufficient permissions.',
  metadata: Metadata { options: undefined, internalRepr: Map {} },
  note: 'Exception occurred in retry method that was not classified as transient' }
```

@schmidt-sebastian do you think this is related to this issue? If not I will open a new one since I can't use the firebase functions right now.

**Comment by schmidt-sebastian** [comment]
@florianlenz I don't think so since you are seeing error code 7. I would check the permissions on your Service Account.

**Comment by schmidt-sebastian** [comment]
For everyone: One of our engineers has narrowed down a behavior change between GRPC and GRPC-JS, which may be related to the issue you are seeing. We don't have a fix yet and will keep this issue updated.

**Comment by sambecker** [comment]
@schmidt-sebastian for what it’s worth I haven’t seen these errors in the 3-4 days since switching to `firebase-admin 7.4.0`. Only that strange document doesn’t exist error I saw in one of my `onCreate` triggers.

**Comment by damienix** [comment]
Similarly to @sambecker, downgrading to `firebase-admin@7.4.8` and `firebase-functions@2.3.1` and I'm having 0 errors for the last 3 days. Thanks for the workaround @schmidt-sebastian 👍

**Comment by levantainteractive** [comment]
This error happens for me on all calls to one of my functions...I'm going to downgrade and see if that helps.

**Comment by TrustyTechSG** [comment]
This is really a big problems for us, because "Cloud Functions guarantees at-least-once execution of a background function for each event emitted by an event source." is not true anymore, and we built a business logic which is fully depends on it....

**Comment by TJHO-capco** [comment]
+1

**Comment by schmidt-sebastian** [comment]
Version 2.2.9 of Firestore might help alleviate some of the pain as we now classify the error you are seeing as retryable.

**Comment by dgudlek** [comment]
Thank you all for your effort to report and investigate this problem.

We are hitting the same issue but from a different client - @google_cloud/datastore.
The exception logs "code: 13" and "method that was not classified as transient". Stack trace looks the same.

Does it make sense to apply the same hotfix as
https://github.com/googleapis/nodejs-firestore/commit/a22ac248b67d9c469af468f1f11e38ff9232dcb9
to
[datastore_client_config.json](https://github.com/googleapis/nodejs-datastore/blob/45bb889327aa278ff71d1811d7bb4bbae40feba8/src/v1/datastore_client_config.json#L7
)
?

**Comment by swftvsn** [comment]
This hits us hard too - the great news is that as we have been around since the betas, we concluded early, that all of cloud function triggers must be run through work queue, that is, for each function invocation, we write a secondary document, that we listen to, which we delete after the function run is done. 

This way to check if all the functions have successfully run, one can just check the work queue if it has any documents. If it has -> update the document and it is rerun.

Any other way has proven to be too unreliable so far.

**Comment by david-arteaga** [comment]
@swftvsn would you mind elaborating a little more on this approach as a workaround?
We currently have `onCreate` and `onUpdate` triggers for docs and we have come across this error for the first time on both triggers.

In the approach you mentioned, would you write the secondary document from the client when you create/update the doc that is supposed to kick off the trigger?

**Comment by swftvsn** [comment]
@david-arteaga Yes, we write always 2 docs using batch, be it an update or create, if we need to invoke cloud function. One is the actual doc, the second one is the work queue entry. Cloud Function triggers are never attached to the main document, always only on the work queue doc. The Cloud Function deletes the work queue entry once it is complete.

As we have some cloud functions doing heavy stuff, we also use this method to build work chains: for an example when new invoice is created a work queue entry is placed:
```
// firestore path: work-queue/${clientId}/send-invoice-start/${workQueueId}
{
  invoicePath: 'clients/clientIdOne/invoices/invoiceId',
  method: 'electronic'
}
```
The first part only checks couple of things, adds counters etc. Once this is ready the work queue entry is deleted from that location, and moved to the next in a batch. The next part is to create the PDF out of the textual invoice data:
```
// firestore path: work-queue/${clientId}/send-invoice-create-pdf/${workQueueId}
{
  invoicePath: 'clients/clientIdOne/invoices/invoiceId',
  method: 'electronic'
}
```
After that is ready, the work queue entry is next moved according to the sending gateway:
```
// firestore path: work-queue/${clientId}/send-invoice-gateway-email/${workQueueId} OR
// firestore path: work-queue/${clientId}/send-invoice-gateway-electronic/${workQueueId} OR
// firestore path: work-queue/${clientId}/send-invoice-gateway-print/${workQueueId} OR
// firestore path: work-queue/${clientId}/send-invoice-gateway-snailmail/${workQueueId} OR
// ...
{
  invoicePath: 'clients/clientIdOne/invoices/invoiceId',
  method: 'electronic'
}
```

This enables us to also monitor which of our integration partners have trouble processing our sends. As all functions are idempodent, we can retry as many times we need and we always have log there to see if Firestore trigger silently is not called. And that happens. Or atleast, we don't have any errors in logs, and we see work queue entries that are never touched before manual intervention.

And finally: **Firestore is GA**, **Cloud Functions are GA** but **Cloud Function Firestore triggers are still in BETA**. So this is to be expected, but we all have to learn it the hard way.

At least I missed it for a long time, that those are in beta and as such may be changed and have bugs. See https://firebase.google.com/docs/functions/firestore-events and notice the beta markings there.

**Comment by wilhuff** [comment]
If you've been affected by this issue, could you please confirm that `@google-cloud/firestore` 2.2.9 addresses it?

**Comment by yuenlye** [comment]
Our previous setup was
```
"firebase-admin": "8.2.0",
"firebase-functions": "^3.0.2"
```

After downgrading to the following on Aug 26, we haven't seen any grpc issue.
```
"firebase-admin": "7.4.0",
"firebase-functions": "2.3.1"
```

**Comment by angeria** [comment]
I haven't seen the issue after upgrade to 2.2.9.

**Comment by sambecker** [comment]
@angeria do you import `firestore 2.2.9` separately? Or is that dependency included with a specific version of `firebase-admin`?

**Comment by angeria** [comment]
@sambecker 
Included with the latest version of `firebase-admin`.

**Comment by h-ARTS** [comment]
@schmidt-sebastian 
I'm getting a similar error. Don't know if this is also related to the same core issue with *grpc*. It throws *error code 5*
```
error: { Error: No document to update: projects/socialpage-3883d/databases/(default)/documents/users/Paki
    at Http2CallStream.call.on (/Users/hananmufti/Desktop/Projects/REACT_APPS/socialapp-functions/functions/node_modules/@grpc/grpc-js/build/src/client.js:96:45)
    at Http2CallStream.emit (events.js:187:15)
    at process.nextTick (/Users/hananmufti/Desktop/Projects/REACT_APPS/socialapp-functions/functions/node_modules/@grpc/grpc-js/build/src/call-stream.js:71:22)
    at process._tickCallback (internal/process/next_tick.js:61:11)
  code: 5,
  details:
   'No document to update: projects/socialpage-3883d/databases/(default)/documents/users/Paki',
  metadata: Metadata { options: undefined, internalRepr: Map {} } }
```

**Comment by schmidt-sebastian** [comment]
@h-ARTS Your error message indicates that you are using `update()` a non-existing document (at least at the time). `update()` verifies that the underlying document exists. You can use `set(..., {merge.true})` if you want to blindly overwrite parts of a document. If no document exists, `set()` will create it.

Note that there are semantic differences between set() and update(). An update call to `foo.bar` will update field `{ foo: { bar: ... }` whereas set() updates the literal field `foo.bar`.

**Comment by h-ARTS** [comment]
@schmidt-sebastian I'm encountering this particular error even when my document was existing.
In this example, the /users/Paki is existing and still, I can't update it. I'm using .update because I want to update only one property value and not everything. But why would I use .set(..., {merge: true}) over .update()?

**Comment by schmidt-sebastian** [comment]
@h-ARTS Do you mind creating a new GitHub issue? This particular issue already has a lot of activity and I suspect that your issue is unrelated. I can help you over there.

If the new issue, can you include for a documentReference.get() and a documentReference.update() call for that particular document? Thanks!

**Comment by h-ARTS** [comment]
@schmidt-sebastian #555

**Comment by slinkstep** [comment]
Having a similar issue with a background function ( onUpdate) that reads and write a good deal of docs (+100): 

Error: Deadline exceeded at Http2CallStream.call.on (/srv/node_modules/@grpc/grpc-js/build/src/client.js:96:45) at emitOne (events.js:121:20) at Http2CallStream.emit (events.js:211:7) at process.nextTick (/srv/node_modules/@grpc/grpc-js/build/src/call-stream.js:71:22) at _combinedTickCallback (internal/process/next_tick.js:132:7) at process._tickDomainCallback (internal/process/next_tick.js:219:9)

**Comment by swftvsn** [comment]
We retry automaticly certain types if errors. For others, there's a timed run to compile an email daily about the numbers of stuck jobs and the ability to bulk restart jobs in our own admin software.

**Comment by swftvsn** [comment]
After upgrading to latest firebase, we're hit with this: https://github.com/grpc/grpc-node/issues/1027 - which is direct result of fixing this issue.

@schmidt-sebastian Why on earth does this library, which is production quality, suddenly start depending on a lib that states "**This library is currently incomplete and experimental**"? I think if you do that, the 3.x line should be marked as experimental too, right?

**Comment by adamduren** [comment]
This issue has wreaked havoc on our application. The firebase team should be communicating this issue to project owners vs owners having to find this thread on their own.

Until a permanent fix is found it seems like the change should be reverted so that people do not upgrade and find out about this the hard way. Especially since it was acknowledged [here](https://github.com/firebase/firebase-functions/issues/536#issuecomment-517461545) that: 
> Our backend team believes that they know what the root cause is, **but it might take quite a while for the issue to be fixed in all production environments.**

**Comment by schmidt-sebastian** [comment]
@swftvsn We made the decision in regards to versioning based on the functionality that we invoke in `@grpc/grpc-js`. We believe (or at least believed at the time) that the subset of the functionality that we rely upon meets our exceptions, as well as the expectations that we conveyed by including in a production-ready library. We understand now that we are not always meeting this bar, and hope that we can close the gap soon.

As you are probably all aware, the GRPC layer we used before was not without issues either. It frequently failed to install (since it required either prebuilt binary or compilation from source) and caused significant startup delays in GCF. On top of that, it's sheer size made it impossible to be used on some environments that are package-size constrained. With that said, I am hesitant to downgrade our dependency for all of our customers.
 
That being said, if the old networking layer works for you, you can still upgrade to Firebase Admin 8 and override the network stack:


```
const admin = require("firebase-admin");
const grpc = require("grpc");

admin.initializeApp();

const firestore = admin.firestore();
firestore.settings({grpc});
```

**Comment by swftvsn** [comment]
@schmidt-sebastian Thank you for the explanation and instructions. Sounds like there are no easy ways out of this right now; I hope that you get it sorted for good and we can all move on.

And still, as always, thank you for the excellent stack!

**Comment by sambecker** [comment]
@schmidt-sebastian do you happen to know if `firebase-admin 8.6.0`, released today, addresses any of this? Or is the version of `firebase-admin` irrelevant if it's not making use of the `grpc` networking layer? Are there plans to make the new networking layer more stable? Has it possibly gotten more stable since this thread began? I'm afraid to test it in production. How can we better follow this issue?

**Comment by mikegamaroff** [comment]
Same problem. Wreaking havoc. 
Error
    at Http2CallStream.call.on (/srv/node_modules/@grpc/grpc-js/build/src/client.js:96:45)
    at emitOne (events.js:121:20)
    at Http2CallStream.emit (events.js:211:7)
    at process.nextTick (/srv/node_modules/@grpc/grpc-js/build/src/call-stream.js:71:22)
    at _combinedTickCallback (internal/process/next_tick.js:132:7)
    at process._tickDomainCallback (internal/process/next_tick.js:219:9)

**Comment by dinvlad** [comment]
Our team decided to implement a general retry mechanism, based (incidentally) on Firestore, to address this and other reliability issues in our background functions.

1.  We [enable retries](https://firebase.google.com/docs/functions/retries) for the functions.
2.  The body of every function goes inside a "catch all" wrapper, described in the following steps.
3.  If the time since `event.timestamp` is past a certain threshold, we log the occasion and "successfully" return from the function early, to prevent indefinite retries. Otherwise, continue.
4.  Similarly to what @swftvsn suggested above, we have a special "event" collection for background functions.
5.  In a transaction, we check if a doc with `eventId` exists in this collection.
6.  If the doc exists, we check the `status` property of that doc:
    - If `status === 'running'`, that means another invocation of the function for this event is already running, so we "successfully" return from the function early, to avoid an extra retry.
    - If `status === 'failed'`, that means a previous invocation failed, so we continue function execution (i.e. a retry in this case).
7. If the doc doesn't exist, we create it (within the same transaction) with `status: 'running'`.
8. Function implementation runs after this transaction, either successfully or with an error.
9. In another transaction, we set `status: 'failed'` if there was an error, and exit with that error.
10. Otherwise, delete the doc with `eventId` in this transaction, and exit successfully.
11. We also _retry this transaction unconditionally_, if there were any Firestore errors during it (otherwise, the function could "fail" with `status` still set to `running`, preventing further retries).

As a result, if the function execution "fails" at any point, incl. its _onEvent_ hook or implementation, it should be retried by GCF runtime automatically. As another bonus, we get the full record of all "permanently failed" invocations in the "event" collection (and with only a short-term storage of "successful" events). This setup also prevents "double-triggering" of function implementation for the same `eventId`. Still, the implementation must be idempotent within a single invocation, to enable safe retries.

**Comment by mikegamaroff** [comment]
Does this mean it was fixed? I'm still getting this error on the creation of new accounts: 

Error
    at Http2CallStream.call.on (/srv/node_modules/@grpc/grpc-js/build/src/client.js:96:45)
    at emitOne (events.js:121:20)
    at Http2CallStream.emit (events.js:211:7)
    at process.nextTick (/srv/node_modules/@grpc/grpc-js/build/src/call-stream.js:71:22)
    at _combinedTickCallback (internal/process/next_tick.js:132:7)
    at process._tickDomainCallback (internal/process/next_tick.js:219:9)

running 
   "firebase-admin": "^8.0.0",
    "firebase-functions": "^3.1.0",

Is an upgrade needed?

**Comment by maxperry** [comment]
The same error also occurred to me once on Sep 21 from a Firestore update trigger:

```
Transaction failed: { Error
at Http2CallStream.call.on (/srv/functions/node_modules/@grpc/grpc-js/build/src/client.js:96:45)
at Http2CallStream.emit (events.js:194:15)
at Http2CallStream.EventEmitter.emit (domain.js:459:23)
at process.nextTick (/srv/functions/node_modules/@grpc/grpc-js/build/src/call-stream.js:71:22)
at process._tickCallback (internal/process/next_tick.js:61:11)
code: 13,
details: '',
metadata: Metadata { options: undefined, internalRepr: Map {} },
note: 'Exception occurred in retry method that was not classified as transient' }
```

I'm running:
```
"firebase-admin": "^8.3.0",
"firebase-functions": "^3.2.0",
```
Does an upgrade fix this issue?

**Comment by iwanether** [comment]
Seems this error happens on coldstart and disappears afterwards until some time passes. It's severely impacting service and I'm recently seeing it pop-up on functions that never exhibited it before. Hope there is a fix soon 🙏

`
at Http2CallStream.call.on (/srv/functions/node_modules/@grpc/grpc-js/build/src/client.js:96:45)
at Http2CallStream.emit (events.js:194:15)
at Http2CallStream.EventEmitter.emit (domain.js:459:23)
at process.nextTick (/srv/functions/node_modules/@grpc/grpc-js/build/src/call-stream.js:71:22)
at process._tickCallback (internal/process/next_tick.js:61:11)
Function execution took 953 ms, finished with status: 'error'
`

`
"node":"10",
"firebase-admin": "^8.6.0",
"firebase-functions": "^3.2.0",`

**Comment by kirillgroshkov** [comment]
Same error happens to us in production (~2 times a day) on a recent (as of today) version of Datastore client.

![image](https://user-images.githubusercontent.com/4757745/65430394-256d1d00-de18-11e9-9947-c9362d65591b.png)

Don't know if it matters, but last 2 times it happened on a sharp times of 16:00 and 22:00 CEST (in case it matters).

Why it is a very BAD error for us is that it causes *unhandledRejection*, despite us having `try {} catch {}` around async function (which supposed to catch both sync and async errors). So, unhandledRejection effectively kills our server (until it restarts in ~15 seconds).

**Comment by swftvsn** [comment]
@dinvlad Thank you for sharing your solution! I think we want to move to similar setup in the near future.

**Comment by lpellegr** [comment]
We have the exact same problem as the original report! Any updates regarding this production issue? 

The Firebase Status Dashboard shows no problems... other than the fact this issue can make businesses lose some leads since their account creation may not be completed appropriately, is nobody shocked by the fact that we are billed for functions failing due to a Firebase internal problem?

**Comment by bottleneck-admin** [comment]
I haven't had an error since the 7th of Sept

**Comment by lpellegr** [comment]
@bottleneck-admin What versions are you using?

**Comment by pot8os** [comment]
Hi @thechenky, any updates? 

I'm facing the same error as I use `datastore` (not `firestore`) right now and I'd like to know the last stable version that is still available in order to do a rollback. It's very severe and has huge impact on me.

<img width="1478" alt="Screen Shot 2019-10-04 at 15 38 42" src="https://user-images.githubusercontent.com/5249113/66186521-663d1f80-e6bd-11e9-9268-95fadf1c569b.png">

My code is as below. I'm sure it had been worked until I updated dependencies.

```typescript
import { Datastore } from '@google-cloud/datastore'
const ds = new Datastore({})

export function saveUserData(id: number, user_data: Data) {
    const key = ds.key([KIND_USER_DATA, ds.int(id)])
    const entity = {
        key: key,
        data: user_data
    }
    ds.save(entity, (err) => {
        if (err != null) {
            console.error(err) // Every invocation gets here after I updated dependencies !!!
        } else {
            console.log("Saved data: " + key.path)
        }
    })
}
```
package.json
```
  "dependencies": {
    "@firebase/app": "^0.4.14",
    "@firebase/app-types": "^0.4.3",
    "@google-cloud/datastore": "^4.2.0",
    "express": "4.16.4",
    "firebase-admin": "^8.3.0",
    "firebase-functions": "^3.2.0"
  },
```

**Comment by pot8os** [comment]
Hmm thanks to explanation by @jakeleventhal at https://github.com/googleapis/nodejs-firestore/issues/768, I understood there is a root cause in `@grpc/grpc-js` lib, and I reverted `@google-cloud/datastore` to 3.x.x that doesn't have it to see how it works. It seems working so far...

**Comment by maxperry** [comment]
Any updates on this issue? The error occurred again a couple times in the last couple hours in one of my Firestore triggers.

**Comment by schmidt-sebastian** [comment]
We are continuing to improve the libraries used in our network stack, but you might still experience occasional network issues even with the latest `@grpc/grpc-js` module.

**Comment by HathTech** [comment]
Any update? 
I am facing the same issue.

**Comment by envyram** [comment]
I had the same issue today as well. We just dodged a bullet on prod issue because of this issue. Has there been an update please? Looks like this has been running for a while

**Comment by piavgh** [comment]
I have the same issue when exporting Firestore to Google Cloud Storage to backup.

It happened randomly for some days, but other days it's ok

**Comment by dinvlad** [comment]
FWIW, we haven't had a prod failure since implementing retries (i.e. we had several "events" with this error, but they were successfully retried). I think in general it's a good idea to implement retries anyways, because network calls are inherently unreliable (thought the library could probably retry them automatically, depending on how easy/safe that is).

**Comment by iamnels1** [comment]
> FWIW, we haven't had a prod failure since implementing retries (i.e. we had several "events" with this error, but they were successfully retried). I think in general it's a good idea to implement retries outside of the library code in any case, because network calls are inherently unreliable (though the library could probably retry these particular errors automatically too, depending on how easy/safe that is).

Hi, thank you for sharing this solution, I also have the same problem with several functions and it is really very problematic, so, as your solution seems to work I will try to implement it; and, rather than reinventing the wheel, May I ask you:  could you give me a little more detail, possibly a skeleton of one of your functions? For example when you say 'The body of every function goes inside a" catch all "wrapper', can I ask you, how do you do that in practice? Thank you in advance for your help; very best.

**Comment by dinvlad** [comment]
@iamnels1 @swftvsn sorry for the delay - here's our example implementation for retries with exponential backoff and jitter: https://gist.github.com/dinvlad/a280c5a44165b24960b3442e5205ab30

**Comment by iamnels1** [comment]
> @iamnels1 @swftvsn sorry for the delay - here's our example implementation for retries with exponential backoff and jitter: https://gist.github.com/dinvlad/a280c5a44165b24960b3442e5205ab30

@dinvlad Really thank you, meanwhile I had set up something that approached that, but with great difficulty and for a less satisfactory result; if you allow it I will use your solution :D

**Comment by sambecker** [comment]
Does anyone know if `firebase-admin 8.8.0` addresses any of this? Or is the version of `firebase-admin` irrelevant if it's not making use of the `grpc` networking layer?

**Comment by madmacc** [comment]
Also having the same issue for the last 3 months.  Last occurring 20 mins ago.  It is intermittent and occurs maybe once a week or so.

**Comment by sambecker** [comment]
What version of `firebase-admin` are you running @madmacc?

**Comment by madmacc** [comment]
@sambecker @google-cloud/firestore": "^2.2.4",
    "firebase-admin": "^8.0.0",
    "firebase-functions": "^3.0.0"

**Comment by madmacc** [comment]
I think there are probably a lot more developers experiencing this issue as it took me a couple of months to even find this thread.
For me if the error is thrown then it appears in the cloud functions log without the "note" portion.
```note: 'Exception occurred in retry method that was not classified as transient' }```
This makes it very difficult to search on.  
Only when I caught the error and logged it did I finally see that message.

**Comment by sambecker** [comment]
Agreed @madmacc I'd be curious to know if upgrading to `firebase-admin 8.8.0` makes any kind of difference. @schmidt-sebastian do you think that would matter?

**Comment by dibikhin** [comment]
After updating all functions I don't see any `'Exception occurred in retry...` for two weeks in to 2 diff projects. I've updated deps and re-deployed the functions. Libs' versions: "firebase-admin": "^8.7.0" and  "firebase-functions": "^3.3.0".

**Comment by schmidt-sebastian** [comment]
@sambecker We have fixed some memory leaks in our networking layer, so I would recommend re-installing firebase-admin to get the newest dependences. You don't actually need to upgrade Firebase Admin or Firastore - a re-install of an older version should also install the newest version of the `@grpc/grpc-js` (via our range dependency on `google-gax`).

We haven't changed client side timeouts or retry behavior, but ware making constant improvements to our backend infrastructure. This might explain the change in error rate.

**Comment by sambecker** [comment]
@schmidt-sebastian is there a way to rebuild dependencies through the `firebase` tool (without updating `package.json`)? Am I understanding correctly that there might still be stability benefits to remaining on `firebase-admin 7`?

**Comment by schmidt-sebastian** [comment]
If you are on firebase-admin 8, removing `node-modules` and re-installing your dependdencies should bring in `google-gax/1.11.1` which is our latest version.

firebase-admin 7 uses a compiled networking layer (rather than a JavaScript implementation). This might work better in some cases, but it does increase your bundle size by 50 MB+, increase function start up times and may even cause installation problems in some environments. If these issues are not a deterrent, you can use firebase-admin 7 directly, or you can opt into the legacy networking layer with firebase-admin 8:

```
const grpc = require('grpc');
const firestore = admin.firestore();
firestore.settings({ grpc });
```

**Comment by madmacc** [comment]
To upgrade to `google-gax to 1.11.1` removing node-modules and re-installing didn't work for me nor did updating `firebase-admin` to the latest version I had to update `@google-cloud/firestore ` to the latest.

**Comment by madmacc** [comment]
Just be careful if you upgrade I seem to have broken my function.  Even after downgrading again I still have `google-gax/1.11.1` and I still have the error.  Not quite sure how to revert back to the previous version of `google-gax`  even if that is what the problem is.

**Comment by adamduren** [comment]
@madmacc if you are using version control and have a .lock file you should be able to get the same set of packages and their dependencies by checking out the version before the upgrade and running `npm ci`. Then run a build of your functions and deploy again. If that doesn't work it sounds like an error unrelated to the upgrade.

Hope that is helpful.

**Comment by hkchakladar** [comment]
I'm getting this issue randomly, 1-2 times in a week.

`Refund error { Error at Http2CallStream.call.on (/srv/node_modules/@grpc/grpc-js/build/src/client.js:101:45) at emitOne (events.js:121:20) at Http2CallStream.emit (events.js:211:7) at process.nextTick (/srv/node_modules/@grpc/grpc-js/build/src/call-stream.js:71:22) at _combinedTickCallback (internal/process/next_tick.js:132:7) at process._tickDomainCallback (internal/process/next_tick.js:219:9) code: 13, details: '', metadata: Metadata { options: undefined, internalRepr: Map {} }, note: 'Exception occurred in retry method that was not classified as transient' }`

which is triggered by `firestore().onUpdate()` and it does usage `transactions` inside.

**Comment by madmacc** [comment]
If anyone actually notices this resolve with the latest version of firebase-admin please let us know in this thread so we can try the same.  I am still getting the error, my aggregates are all over the show.

**Comment by sambecker** [comment]
Just upgraded from `firebase-admin 7.x` to `firebase-admin-8.8.0` and have seen one of these mysterious crashes in 1-2 weeks of use. So it's possibly better, but still unreliable? This is my most recent error, referenced in issue https://github.com/firebase/firebase-functions/issues/522:
```
Error: 13 INTERNAL:      at callErrorFromStatus (/srv/node_modules/@grpc/grpc
js/build/src/call.js:30:26)     at Http2CallStream.call.on (/srv/node_modules/@grpc/grpc
js/build/src/call.js:79:34)     at emitOne (events.js:121:20)     at Http2CallStream.emit
(events.js:211:7)     at process.nextTick (/srv/node_modules/@grpc/grpc-js/build/src/call
stream.js:75:22)     at _combinedTickCallback (internal/process/next_tick.js:132:7)     at
process._tickDomainCallback (internal/process/next_tick.js:219:9)
```
FWIW this was seen in a once-a-day scheduled cloud function that worked correctly the other six days of the week.

@schmidt-sebastian would this suggest to you that I should opt into the `grpc` legacy networking layer? I know I've asked before but can you think of any variables that might make the newer, uncompiled Javascript networking layer more stable? What tends to blow it up? Cold starts? Too many promises? Too many `firestore` updates/network requests? How can I make my app more resilient to these crashes?

**Comment by madmacc** [comment]
Is this only happening for a subset of projects/developers?  
I don't understand why this is not a bigger issue for people.
This is still happening every 2-3 days for me.
I am just about to start another project that requires accurate aggregation for availability and I am not confident this can work in firestore.  

Thanks for the update @sambecker seems like there is not a lot of point in me upgrading yet.

**Comment by alqamabinsadiq** [comment]
I am also facing the same issue it happens randomly after 1-2 days. I am facing this issue while updating the document in Firestone.

**Comment by katopz** [comment]
FYI : I will note my experience here for record.

What may causing error code 13
- Cold start
- `onCall`
- Batch set with `FieldValue.increment`
```js
  batch.set(
    ref,
    {
      rating: FieldValue.increment(10),
      rated: FieldValue.increment(1)
    },
    { merge: true }
  )
```

- a lot of async/await
- some amount of import

Work around
- Pre-call next function to warm it up.
- Avoid use `FieldValue.increment` // Do normal read and write as possible.
- Don't use async/await that much.
- Watch out for unnecessary import from other file with have unrelated import.

I still facing this on local develop on latest release and really hope it will get better on  prod ;)

**Comment by mikron123** [comment]
Why do I need to "warm up" a function?! What is it a car engine?

**Comment by dinvlad** [comment]
Just my 2c, we've been using these for several months now without any problems once I implemented proper retries (as described above).

**Comment by dgobaud** [comment]
we are seeing the problem too periodically - fix would be great

    "firebase-admin": "^8.4.0",
    "firebase-functions": "^3.2.0",

**Comment by dgobaud** [comment]
@thechenky any update? keeps happening :(

**Comment by GideonWislang** [comment]
Just came across this error today was working fine yesterday :(

**Comment by jbromberg** [comment]
Also having this error. Major issue that is causing problems in my application.

**Comment by samstr** [comment]
This has suddenly started appearing for me in the past hour.

**Comment by alqamabinsadiq** [comment]
It's happening for everyone I think.

**Comment by sambecker** [comment]
FWIW I haven't seen errors like these since upgrading to Node 10, which I understand is still in beta:

```
"engines": {
  "node": "10"
}
```

**Comment by dgobaud** [comment]
i saw this in this package https://www.npmjs.com/package/@google-cloud/pubsub google told me update to latest it fixed it i was using really old version. havent seen it since

**Comment by jbromberg** [comment]
I haven’t seen these errors until I upgraded to Node 10.

On Jun 23, 2020, at 12:57 PM, Sam Becker <notifications@github.com> wrote:

﻿

FWIW I haven't seen errors like these since upgrading to Node 10, which I understand is still in beta:

"engines": {
  "node": "10"
}

**Comment by dinvlad** [comment]
Yes, I think we've seen them both in PubSub and Firestore with at least Node.js 10. I can only conclude that they're either not caught properly by the GRPC library, or it's virtually impossible to protect against all of sporadic errors like these, so the strategy we've successfully adopted is just implementing proper retries ourselves, as noted above.

**Comment by schmidt-sebastian** [comment]
@bencoe for visibility.

**Comment by dzoba** [comment]
I started hitting this error today.  Not even on a function, just on a nodejs script with firebase-admin I was running against firestore to do some DB updates.

**Comment by samstr** [comment]
For me I was seeing this error as well as "No connection" and "Deadline exceeded" when performing large numbers of writes individually. Refactoring to use [batched writes](https://firebase.google.com/docs/firestore/manage-data/transactions#batched-writes) solved it for me.

**Comment by manwithsteelnerves** [comment]
We got this error today
"Exception occurred in retry method that was not classified as transient"

**Comment by damienromito** [comment]
I have the same probleme on differents endpoints.
Here are the logs if it helps : 
![Screen Shot 2020-07-27 at 16 26 38](https://user-images.githubusercontent.com/603808/88553788-f6ba6700-d025-11ea-8729-f4ca72281a8b.png)

![Screen Shot 2020-07-27 at 16 27 04](https://user-images.githubusercontent.com/603808/88553864-0e91eb00-d026-11ea-984d-2acd78b44bb9.png)

**Comment by UnJavaScripter** [comment]
Everything was woking fine with my functions with the emulator for several hours and suddenly I got that same error:
```
{
  code: 2,
  details: '',
  metadata: Metadata {
    internalRepr: Map(1) { 'content-type' => [Array] },
    options: {}
  },
  note: 'Exception occurred in retry method that was not classified as transient'
}
```

Using node 14 without any problem. Tried using 10 and nothing changed.

I've been using the Functions emulator with Firestore triggers. The data will be written regardless, but no logs are shown (and also the error we're all having is thrown).

**Comment by kythin** [comment]
This error was driving me nuts, but it turned out to be some form of permission thing for me. After updating the service account to just have 'Project -> Editor' access the firestore writes started working again.

Obviously not ideal to give the service account such wide access, but it's a start!

**Comment by jonrandahl** [comment]
Just adding my own two pence as I've landed on this thread too many times not to now... 

I can concur with @manwithsteelnerves in that adding the `Pub/Sub Editor` role to the service account has somehow stopped the DEADLINE EXCEEDED errors occurring in one of our cloud functions on one of our instances.

However, on another of our instances where the Service account does not have the `Pub/Sub Editor` role, the same cloud function is working fine vie both the local emulator and deployed to that same non-local instance which the emulator was connected to, and the cloud function does not error on that instance.

Thanks to everyone nonetheless for all your hard work to resolve this issue, and for all the comments that have helped others including myself previously!

Please let me know if you would like further information. 🙏

**Comment by ebeloded** [comment]
I can reliably get this exception when setting an environment variable SERVER_ENV='local'. Removing the var or setting it to a value other than 'local' makes the error go away

**Comment by fabiank0** [comment]
Got it today while handling files that were in total over 3MB. No Triggers are used. 

Error: 14 UNAVAILABLE: No connection established at Object.callErrorFromStatus (/workspace/node_modules/@grpc/grpc-js/build/src/call.js:30:26) at Object.onReceiveStatus (/workspace/node_modules/@grpc/grpc-js/build/src/client.js:175:52) at Object.onReceiveStatus (/workspace/node_modules/@grpc/grpc-js/build/src/client-interceptors.js:341:141) at Object.onReceiveStatus (/workspace/node_modules/@grpc/grpc-js/build/src/client-interceptors.js:304:181) at Http2CallStream.outputStatus (/workspace/node_modules/@grpc/grpc-js/build/src/call-stream.js:116:74) at Http2CallStream.maybeOutputStatus (/workspace/node_modules/@grpc/grpc-js/build/src/call-stream.js:155:22) at Http2CallStream.endCall (/workspace/node_modules/@grpc/grpc-js/build/src/call-stream.js:141:18) at Http2CallStream.cancelWithStatus (/workspace/node_modules/@grpc/grpc-js/build/src/call-stream.js:457:14) at ChannelImplementation.tryPick (/workspace/node_modules/@grpc/grpc-js/build/src/channel.js:237:32) at Object.updateState (/workspace/node_modules/@grpc/grpc-js/build/src/channel.js:106:26)

Node 10, 
    "@google-cloud/storage": "^4.3.1",
    "firebase-admin": "^8.9.2",
    "firebase-functions": "^3.8.0",

**Comment by Pixelwelder** [comment]
I cannot tell you how many days I burned on this issue. In my case, I left a curly bracket out of a document address.

`.document('a/{docA}/chat/{docB') // missing final bracket`

The error was occurring in a totally different function.

**Comment by MattGoldwater** [comment]
Is anyone else still getting this error?

**Comment by Dric0** [comment]
> Is anyone else still getting this error?

I did, yesterday. But I only get these once every 2 months

**Comment by MattGoldwater** [comment]
> > Is anyone else still getting this error?
> 
> I did, yesterday. But I only get these once every 2 months

Thanks for answering. Yeah it turned out I got it because I had a syntax error I wasn't aware of.

**Comment by calclavia** [comment]
I'm getting the same issue when I have to write a lot of documents in many async calls (500+ async writes).

**Comment by inlined** [comment]
Hi there, it seems like the original bug has been fixed and we're getting more reports of possibly more than one bug. Collecting them in the wrong old issue can hurt our ability to triage and get your issues taken care of, so I'm going to close it and let you open new bugs that can be resolved with more specific conversations.

The original bug was due to a networking issue that can happen when a server is idle: the connection gets reset and the next request may fail. Originally this was a problem with the gRPC library because it wasn't handling a clean connection reset. This problem also happens more generally when the FIN packet isn't sent across the internet due to a number of reasons involving performance and security. If the library isn't already aware the connection is invalid (e.g. the FIN packet was dropped or the library isn't handling FIN correctly) the next request will fail. Thanks to the [Two Generals Problem](https://en.wikipedia.org/wiki/Two_Generals%27_Problem), it's impossible to know if a request failed before or after the server got your request. The library can retry if it knows the request is idempotent (e.g. GET) but it can't necessarily retry if the request isn't (e.g. POST). Fortunately, _you_ might know that your code is idempotent. In fact, our guidance is that _all_ cloud functions should be idempotent because you may get more than one invocation. So a retry at the application level should be safe.

Normally you can retry with a simple `try/catch`. Diving through some of this bug and the internal ones as well, it looks like you couldn't always `catch` an error in the gRPC library. If that's (still) the case, it's an issue and someone should file a new bug against [the gRPC repo](https://github.com/grpc/grpc-node) or possibly [gax-nodejs](https://github.com/googleapis/gax-nodejs) that exceptions cannot be caught. A foolproof way to handle exceptions anywhere in your codebase is to turn on retries in your functions. This adds a risk that a crash loop will cause indefinite executions, so you'll need to find some way to drop events-of-death.

I can guarantee you that the event type you're listening to has no impact on this issue. It's happening because your function was idle this whole time and we didn't garbage collect the container so that you could avoid a cold start. Crashes popping up in the Firestore/Datastore library should probably be filed against those SDKs ([nodejs-firestore](https://github.com/googleapis/nodejs-firestore) and [nodejs-datastore](https://github.com/googleapis/nodejs-datastore). If you get an obvious networking error,  you could also consider filing a bug against the gRPC library instead. You can of course file a bug against this repo as a starting point, but you just might have a slower response as we find the right people and move your bug to the right location. You're our customers and we care about your experience; this repo just isn't where the exception lies so it's not where the fix will come.

**Comment by MNorgren** [comment]
> I cannot tell you how many days I burned on this issue. In my case, I left a curly bracket out of a document address.
> 
> `.document('a/{docA}/chat/{docB') // missing final bracket`
> 
> The error was occurring in a totally different function.

Wow! Thanks. I double I would have found this if you hadn't mentioned it!!! In my case, I was using a dollar sign in my path that caused this error... Seriously.

`.document('a/{docA}/chat/${docB'}) `

**Comment by radhikadeo** [comment]
I am still getting this same issue, can someone help?

**Comment by Pixelwelder** [comment]
@radhikadeo Just to verify, you've checked _all_ your paths?

**Comment by moritzmorgenroth** [comment]
I ran into this problem when working with Firestore Point-in-time recovery (PITR) (which is an awesome beta feature! 🎉).

**Comment by hkchakladar** [comment]
The issue arises randomly for me. I'm seeing it once/twice a month (out of ~ 10k in a month).

**Comment by vikasdduc** [comment]
getting error when trying to delete topic and subscription with delete method  anyone can help?                                                                           
 code: 7,
  details: 'User not authorized to perform this action.',
  metadata: Metadata { internalRepr: Map(0) {}, options: {} },
  note: 'Exception occurred in retry method that was not classified as transient'

**Comment by sandhya1349** [comment]
@vikasdduc , can you please let us know the module and node versions?

**Comment by olboghgc** [comment]
same here 

> @taeold Could we please re-open this issue? It appears to still be occurring for users. It happened to me yesterday on node 18 and `firebase-functions@4.6.0`

same here

**Comment by nojaf** [comment]
Got this today with `node v20.11.0` and `firebase-functions 5.0.1`

```
at async runHTTPS (/home/nojaf/.bun/install/global/node_modules/firebase-tools/lib/emulator/functionsEmulatorRuntime.js:531:5)\n    at async /home/nojaf/.bun/install/global/node_modules/firebase-tools/lib/emulator/functionsEmulatorRuntime.js:694:21 {\n  code: 2,\n  details: '',\n  metadata: Metadata {\n    internalRepr: Map(1) { 'content-type' => [Array] },\n    options: {}\n  },\n  note: 'Exception occurred in retry method that was not classified as transient'\n}"}
```

**Comment by haayhappen** [comment]
Still happening to me as well cc @schmidt-sebastian

**Comment by dschnare** [comment]
We are getting this error as well. It was during a WriteBatch.commit when we were updating one document. We have been using @google-cloud/firestore@6.6.0.

**Comment by spock123** [comment]
@dschnare  did you try to use the `preferRest` setting in the configuration?

**Comment by SoufianeBenbah** [comment]
Personally, I have had this problem several times, and each time it has something to do with my code.
Either I forgot a `{` somewhere or something like that.

The last one was because I tried to connect to a database that didn't exist.
```
const db = admin.firestore();
db.settings({ databaseId:"non-existante-db", ignoreUndefinedProperties: true })
```

Hope this helps some of you

**Comment by YagoAyala** [comment]
A lot of years later and I am having the exact same problem

**Comment by cabljac** [comment]
@YagoAyala  can you open a fresh issue for us to track, I suspect you have different dependencies etc. We will try to reproduce and provide a fix.
